passes/cmds/scratchpad.cc

```diff
@@ -0,0 +1,130 @@
+		log("\n");
+		log("The identifier may not contain whitespace. By convention, it is usually prefixed\n");
+		log("by the name of the pass that uses it, e.g. 'opt.did_something'. If the value\n");
+		log("contains whitespace, it must be enclosed in double quotes.\n");
+		log("\n");
+	}
+
+	void execute(std::vector<std::string> args, RTLIL::Design *design) YS_OVERRIDE
+	{
+		size_t argidx;
+		for (argidx = 1; argidx < args.size(); argidx++)
+		{
+			if (args[argidx] == "-get" && argidx+1 < args.size()) {
+				string identifier = args[++argidx];
+				if (design->scratchpad.count(identifier)){
+					log("%s\n", design->scratchpad_get_string(identifier).c_str());
+				} else {
+					log("\"%s\" not set\n", identifier.c_str());
+				}
+				continue;
+			}
+			if (args[argidx] == "-set" && argidx+2 < args.size()) {
+				string identifier = args[++argidx];
+				string value = args[++argidx];
+				if (value.front() == '\"' && value.back() == '\"') value = value.substr(1, value.size() - 2);
+				design->scratchpad_set_string(identifier, value);
+				continue;
+			}
+			if (args[argidx] == "-unset" && argidx+1 < args.size()) {
+				string identifier = args[++argidx];
+				design->scratchpad_unset(identifier);
+				continue;
+			}
+			if (args[argidx] == "-copy" && argidx+2 < args.size()) {
+				string identifier_from = args[++argidx];
+				string identifier_to = args[++argidx];
+				if (design->scratchpad.count(identifier_from) == 0) log_error("\"%s\" not set\n", identifier_from.c_str());
+				string value = design->scratchpad_get_string(identifier_from);
+				design->scratchpad_set_string(identifier_to, value);
+				continue;
+			}
+			if (args[argidx] == "-assert" && argidx+2 < args.size()) {
+				string identifier = args[++argidx];
+				string expected = args[++argidx];
+				if (expected.front() == '\"' && expected.back() == '\"') expected = expected.substr(1, expected.size() - 2);
+				if (design->scratchpad.count(identifier) == 0)
+					log_error("Assertion failed: scratchpad entry '%s' is not defined\n", identifier.c_str());
+				string value = design->scratchpad_get_string(identifier);
+				if (value != expected) {
+					log_error("Assertion failed: scratchpad entry '%s' is set to '%s' instead of the asserted '%s'\n",
+					           identifier.c_str(), value.c_str(), expected.c_str());
+				}
+				continue;
+			}
+			if (args[argidx] == "-assert-set" && argidx+1 < args.size()) {
+				string identifier = args[++argidx];
+				if (design->scratchpad.count(identifier) == 0)
+					log_error("Assertion failed: scratchpad entry '%s' is not defined\n", identifier.c_str());
+				continue;
+			}
+			if (args[argidx] == "-assert-unset" && argidx+1 < args.size()) {
+				string identifier = args[++argidx];
+				if (design->scratchpad.count(identifier) > 0)
+					log_error("Assertion failed: scratchpad entry '%s' is defined\n", identifier.c_str());
+				continue;
+			}
+			log("Unrecognized argument: %s\n", args[argidx].c_str());
```

Comment:
It is typical to fall through to the `extra_args()` like so:
https://github.com/YosysHQ/yosys/blob/a73f96594f4688afc85098b485ef7788e79f5c33/passes/hierarchy/hierarchy.cc#L749-L751
as it produces a nice error message point to exactly what argument was failing and where it was in the command.